Updated log10 calcs to use faster log2 approach

ghostop14 opened the pull request:
Calculating n*log10(x) is faster as:
n*log2(x)/log2(10) = (n/log2(10)) * log2(x)
This update incorporates that change.

Testing before updates gave the following performance (3664 ms -> 270 ms):
RUN_VOLK_TESTS: volk_32fc_s32f_power_spectrum_32f(131071,1987)
a_sse3 completed in 3656.11 ms
generic completed in **3664 ms**
Best aligned arch: a_sse3
Best unaligned arch: generic

After updates:
RUN_VOLK_TESTS: volk_32fc_s32f_power_spectrum_32f(131071,1987)
a_sse3 completed in 2049.2 ms
generic completed in **270.026 ms**
Best aligned arch: generic
Best unaligned arch: generic

ghostop14 commented:
Pushed another update to the spectral kernels to leverage the volk multiply kernel for the one CPU loop I had.  Now only a scalar add has to be done with a standard loop (doesn't look like there's a volk 32f scalar add kernel).  Squeezed a little more performance out of it.

ghostop14 commented:
Looks like there's a build failure in qa_volk_32f_x3_sum_of_poly_32f, unrelated to these changes.

michaelld commented on include/volk/volk_common.h:
```diff
@@ -5,84 +5,86 @@
-#  define __VOLK_PREFETCH(addr)
-#  define __VOLK_ASM __asm
-#  define __VOLK_VOLATILE
+#define __VOLK_ATTR_ALIGNED(x) __declspec(align(x))
```
did you run clang-format?

michaelld replied:
I don't mind; but please make it a separate commit, so that we can easily review the critical changes.

michaelld commented on kernels/volk/volk_32f_log2_32f.h:
```diff
@@ -92,29 +94,23 @@
+static inline void volk_32f_log2_32f_generic(float *bVector,
+                                             const float *aVector,
+                                             unsigned int num_points) {
+  float *bPtr = bVector;
```
i don't know clang-format; like really at all. can we change it to do `float* foo;` rather than `float *foo;`? My preference is keeping the whole type all together.

michaelld reviewed (changes requested):
this is tough to review because of the clang-format changes. Please split them up & redo this PR to make our review doable easily. thx!

argilo commented on kernels/volk/volk_32fc_s32f_power_spectrum_32f.h:
```diff
@@ -128,7 +128,10 @@ volk_32fc_s32f_power_spectrum_32f_a_sse3(float* logPowerOutput, const lv_32fc_t*
 
-    *destPtr = 10.0*log10f(((real * real) + (imag * imag)) + 1e-20);
+    // *destPtr = 10.0*log10f(((real * real) + (imag * imag)) + 1e-20);
+    // For performance rewrite log10 to use log2
```
I don't think these comments are needed. (Likewise for the other places they appear below.)

argilo commented on kernels/volk/volk_32fc_s32f_power_spectrum_32f.h:
```diff
@@ -128,7 +128,10 @@ volk_32fc_s32f_power_spectrum_32f_a_sse3(float* logPowerOutput, const lv_32fc_t*
+    // *destPtr = 10.0*log10f(((real * real) + (imag * imag)) + 1e-20);
+    // For performance rewrite log10 to use log2
+    *destPtr = volk_log2to10factor *
+               log2f_non_ieee(((real * real) + (imag * imag)) + 1e-20);
```
The `+ 1e-20` should no longer be necessary, because `log2f_non_ieee` already returns -127 instead of -Inf when the input is zero.

argilo replied:
There are a bunch more places below where `+ 1e-20` is used and I think they can all be removed.

ghostop14 commented:
New updates have been pushed.  Thanks to @argilo we eliminated a little more unnecessary code from the calculations.

argilo commented:
The entire `volk_32fc_s32f_x2_power_spectral_density_32f` implementation could be replaced by a single call to `volk_32fc_s32f_power_spectrum_32f(logPowerOutput, complexFFTInput, normalizationFactor * sqrt(rbw), num_points)`.

argilo commented:
Since power spectrum can be implemented as four calls to other volk kernels, it hardly seems worth keeping the sse3 and avx kernels, which are an order of magnitude slower.

argilo commented:
I've implemented the above two suggestions in this commit: https://github.com/argilo/volk/commit/a5a23f77b370e3989384302e21a5e9902031c6cb

ghostop14 commented:
Before I merge the commit in, @michaelld, @jdemel: we were talking if it made sense to leave the other non-generic kernels for now to allow tests to run since it was such a big change, then in a future update remove them and consolidate the spectral calls to just an inline function?  Or would it make sense to do this within this PR?   Any suggestions?

michaelld commented:
My vote is to remove those kernels in a different commit.

michaelld commented:
Let's get this issue resolved & see where the speedups happen once it's all said and done. We can then decide what the correct path is moving forward.

ghostop14 commented:
I think we should be good-to-go then.  All tests now passed and I incorporated the volk_32fc_s32f_x2_power_spectral_density_32f into this version that @argilo had in the commit.

ghostop14 commented:
@jdemel , @michaelld , hope all is well with everything going on.  Any updates on these and the other PR's for me to address before merging?  Obviously have a little more spare time than usual in the evenings to clean up any issues given all the closures.

jdemel commented on kernels/volk/volk_32fc_s32f_power_spectrum_32f.h:
```diff
@@ -190,23 +193,25 @@ volk_32fc_s32f_power_spectrum_32f_generic(float* logPowerOutput, const lv_32fc_t
+  // * log2(x)
+  volk_32f_log2_32f(logPowerOutput, logPowerOutput, num_points);
+  volk_32f_s32f_multiply_32f(logPowerOutput, logPowerOutput, volk_log2to10factor,
+                             num_points);
```
I like the "meta" approach here.

Since this is the generic kernel, I wonder if we should call the generic kernels here as well.

ghostop14 replied:
We'd lose all the performance benefits if we forcibly call the generics.  I think the benefit here is that this generic calls the best-performing version of each of the called kernels.

jdemel replied:
That's true. Also, a combination of different kernels for different parts might be beneficial in terms of performance. In this case, I wonder if we should just remove all the other kernels. Or go for a generic simple implemenentation and a generic optimized version that just calls into the other kernels.

jdemel reviewed (commented):
LGTM. We should discuss what to do about my comment and then I'm all in to merge this PR.

ghostop14 commented:
I thought about some options too, but this one's kind of a one-off since it's not a foundational kernel per se.  It just calls other kernels.  It's more of an inline wrapper than a kernel?  In that respect I don't know taking the time to write a pure generic to see the implementation would help.  Maybe we just add the implementation calculations back in to a comment to show the logic?

jdemel commented:
@ghostop14 do you mind to rebase your PR? That would be great.

ghostop14 committed:
Updated log10 calcs to use faster log2 approach

Calculating n*log10(x) is faster as:
n*log2(x)/log2(10) = (n/log2(10)) * log2(x)
This update incorporates that change.  The
generic kernels were also streamlined to
take advantage of other volk kernels and let
the dispatcher pick the most optimized kernel
for each step.

ghostop14 commented:
Rebased, fixed merge issues, clang-formatted my modified files, moved generic to top of file to align with other suggestions, and re-pushed.  Doesn't look like CI re-ran after force push but the conflicts are resolved in the repo.

ghostop14 committed:
Merge branch 'master' into log10

ghostop14 committed:
updates

michaelld reviewed (approved):
LGTM

michaelld commented:
@jdemel thoughts on this PR? I'm good with it.

jdemel commented:
I approve this PR too. Merging.
